gh-99813: Start using `SSL_sendfile` when available

illia-v opened the pull request:
<!-- gh-issue-number: gh-99813 -->
* Issue: gh-99813
<!-- /gh-issue-number -->

New `_ssl__SSLSocket_sendfile_impl` function is very similar to `_ssl__SSLSocket_write_impl` with a few adjustments for the [`SSL_sendfile`](https://www.openssl.org/docs/man3.1/man3/SSL_sendfile.html) syscall.

I am glad there was non-SSL [`socket.socket.sendfile`](https://docs.python.org/3/library/socket.html#socket.socket.sendfile) and underlying ``socket.socket._sendfile_use_sendfile`` implemented. I reused that code for new `ssl.SSLSocket._sendfile_use_ssl_sendfile` by moving shared logic to `socket.socket._sendfile_zerocopy`.

And `test.test_ssl.ThreadedTests.test_sendfile` needed to have a socket bound before an SSL context wrapped it because this seemed to affect kTLS availability.

illia-v committed:
Add methods for checking whether kTLS is used

illia-v committed:
Start using `SSL_sendfile` when available

illia-v committed:
Try fixing warnings

illia-v committed:
Use unsigned `size_t` instead of `Py_ssize_t`

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main'

Conflicts:
	Modules/clinic/_ssl.c.h

illia-v committed:
Make `sendfile_impl` more similar to `write_impl`

illia-v committed:
Modify `test_ssl` to test `SSL_sendfile` calls

illia-v committed:
Modify documentation and add a news entry

illia-v committed:
Fix a test

illia-v committed:
Rename `uses_ktls_for_write` to `uses_ktls_for_send`

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Update `versionchanged`

illia-v committed:
Update a test to get successful HTTP responses

illia-v committed:
Add setting errors

illia-v committed:
Fix conditions

illia-v committed:
Add handling of `SSL_R_UNINITIALIZED`

This is similar to https://github.com/nginx/nginx/blob/1c6183725247024f1bca73ac9a833098af7558af/src/event/ngx_event_openssl.c#L3078-L3089

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Modify `test_sendfile` to avoid the internet

illia-v committed:
Refactor `test_sendfile` a bit

gpshead commented on Modules/_ssl.c:
```diff
@@ -2330,6 +2360,170 @@ PySSL_select(PySocketSockObject *s, int writing, _PyTime_t timeout)
+    int uses = BIO_get_ktls_send(SSL_get_wbio(self->ssl));
+    return PyBool_FromLong((long)uses);
+#else
+    return Py_False;
```
Py_INCREF before returning the singleton anywhere you return Py_False or Py_True.

illia-v replied:
Thanks for the feedback! It looks like `Py_INCREF` for boolean objects has not been needed since this commit https://github.com/python/cpython/commit/ea2c0016502472aa8baa3149050ada776d17a009#diff-cb015dd8a8161c3ca18f71fc78aac892d865a2a096a13ffc03a9cda8b02970ae. [Documentation](https://docs.python.org/3.12/c-api/bool.html#c.Py_False) no longer mentions the need too.

picnixz replied:
Instead use `Py_RETURN_FALSE`:

```suggestion
    Py_RETURN_FALSE;
```

gpshead reviewed (commented):
I won't have time to look at this in detail for a while. I'm just dropping a note about one issue seen while quickly skimming it.

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Try to fix new warnings

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Apply a change from ea9a296fce2f786b4cf43c7924e5de01061f27ca

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Set `versionchanged` to next in docs

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

picnixz commented on Lib/socket.py:
```diff
@@ -348,79 +349,86 @@ def makefile(self, mode="r", buffering=None, *,
+                file.seek(offset)
+
+    def _sendfile_use_sendfile(self, file, offset=0, count=None):
+        if not (sendfile := getattr(os, 'sendfile', None)):
```
Please keep the global `if hasattr` guard at the class level. Otherwise this will be executed everytime the method is called.

picnixz commented on Lib/ssl.py:
```diff
@@ -1265,15 +1269,31 @@ def sendall(self, data, flags=0):
+        if self._sslobj.uses_ktls_for_send():
+            try:
+                return self._sendfile_use_ssl_sendfile(file, offset, count)
+            except _GiveupOnSSLSendfile:
```
I would prefer the following:

```py
if not self._sslobj.uses_ktls_for_send():
    return self._sendfile_use_send(file, offset, count)

sendfile = getattr(self._sslobj, "sendfile", None)
if sendfile is None:
    return self._sendfile_use_send(file, offset, count)

try:
    return self._sendfile_zerocopy(
        sendfile, _GiveupOnSSLSendfile, file, offset, count,
    )
except _GiveupOnSSLSendfile:
    return self._sendfile_use_send(file, offset, count)    
```

and remove `_sendfile_use_ssl_sendfile`.

picnixz commented on Lib/socket.py:
```diff
@@ -348,79 +349,86 @@ def makefile(self, mode="r", buffering=None, *,
+            file=file,
+            offset=offset,
+            count=count,
+        )
```
```suggestion
        return self._sendfile_zerocopy(
            partial(os.sendfile, self.fileno()),
            _GiveupOnSendfile,
            file, offset, count,
        )
```

It's a bit easier to read IMO.

picnixz commented on Lib/socket.py:
```diff
@@ -348,79 +349,86 @@ def makefile(self, mode="r", buffering=None, *,
         return text
 
-    if hasattr(os, 'sendfile'):
+    def _sendfile_zerocopy(self, zerocopy_func, giveup_err, file, offset=0, count=None):
```
Let's rename `giveup_err` to `giveup_exc_type` instead.

picnixz commented on Lib/socket.py:
```diff
@@ -348,79 +349,86 @@ def makefile(self, mode="r", buffering=None, *,
         return text
 
-    if hasattr(os, 'sendfile'):
+    def _sendfile_zerocopy(self, zerocopy_func, giveup_err, file, offset=0, count=None):
```
Just to be sure, the implementation didn't change at all, just that it was now more generic right?

illia-v replied:
Yes, it didn't

picnixz commented on Misc/NEWS.d/next/Library/2023-03-13-22-51-40.gh-issue-99813.40TV02.rst:
```diff
@@ -0,0 +1,4 @@
+Python now uses ``SSL_sendifle`` internally when it is possible (see
```
```suggestion
Python now uses ``SSL_sendfile`` internally when it is possible (see
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -75,6 +75,36 @@
 
 
 
```
```suggestion
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -75,6 +75,36 @@
+#endif
+    if (PyErr_Occurred())
+        return 0;
+    return 1;
```
```suggestion
    return PyErr_Occurred() ? 0 : 1;
```

(so that we have an explicit 0/1)

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    int uses = BIO_get_ktls_recv(SSL_get_rbio(self->ssl));
+    return PyBool_FromLong((long)uses);
+#else
+    return Py_False;
```
```suggestion
    Py_RETURN_FALSE;
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+{
+#ifdef BIO_get_ktls_recv
+    int uses = BIO_get_ktls_recv(SSL_get_rbio(self->ssl));
+    return PyBool_FromLong((long)uses);
```
```suggestion
    return PyBool_FromLong(uses);
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    int has_timeout;
+
+    if (sock != NULL) {
+        if (((PyObject*)sock) == Py_None) {
```
```suggestion
        if ((PyObject *)sock == Py_None) {
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+        PyErr_SetString(get_state_sock(self)->PySSLErrorObject,
+                        "Underlying socket too large for select().");
+        goto error;
+    }
```
PEP-7

```suggestion
    if (sockstate == SOCKET_HAS_TIMED_OUT) {
        PyErr_SetString(PyExc_TimeoutError,
                        "The write operation timed out");
        goto error;
    }
    else if (sockstate == SOCKET_HAS_BEEN_CLOSED) {
        PyErr_SetString(get_state_sock(self)->PySSLErrorObject,
                        "Underlying socket has been closed.");
        goto error;
    }
    else if (sockstate == SOCKET_TOO_LARGE_FOR_SELECT) {
        PyErr_SetString(get_state_sock(self)->PySSLErrorObject,
                        "Underlying socket too large for select().");
        goto error;
    }
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+        self->err = err;
+
+        if (PyErr_CheckSignals())
+            goto error;
```
```suggestion
        if (PyErr_CheckSignals()) {
            goto error;
        }
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+             err.ssl == SSL_ERROR_WANT_WRITE);
+
+    if (err.ssl == SSL_ERROR_SSL
+        && ERR_GET_REASON(ERR_peek_error()) == SSL_R_UNINITIALIZED) {
```
```suggestion
        && ERR_GET_REASON(ERR_peek_error()) == SSL_R_UNINITIALIZED)
    {
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    if (retval < 0)
+        return PySSL_SetError(self, __FILE__, __LINE__);
+    if (PySSL_ChainExceptions(self) < 0)
+        return NULL;
```
```suggestion
    if (retval < 0) {
        return PySSL_SetError(self, __FILE__, __LINE__);
    }
    if (PySSL_ChainExceptions(self) < 0) {
        return NULL;
    }
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    return PyLong_FromSize_t(retval);
+error:
+    Py_XDECREF(sock);
+    PySSL_ChainExceptions(self);
```
```suggestion
    (void)PySSL_ChainExceptions(self);
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+
+    if (sock != NULL) {
+        /* just in case the blocking state of the socket has been changed */
+        nonblocking = (sock->sock_timeout >= 0);
```
This can be a local variable. Most of the file uses ANSI C conventions but new code is allowed to declare variables near their usages unless it's easier to put them at the top for maintenance purposes.

illia-v replied:
Thank you for the review! The new function was based on `_ssl__SSLSocket_write_impl`, so I kept style of the existing function, I'll change it per your suggestions

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+            sockstate = SOCKET_OPERATION_OK;
+        }
+
+        if (sockstate == SOCKET_HAS_TIMED_OUT) {
```
You can use a switch instead of chained ifs. It may be a bit easier to read and to PEP-7ify.

illia-v replied:
Not sure about using switch here because `else if (sockstate == SOCKET_IS_NONBLOCKING)` uses `break` to exit the loop. It's not possible to do from switch, is it?

picnixz replied:
Ah. Yes indeed. Lets keep the if

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+            timeout = _PyDeadline_Get(deadline);
+        }
+
+        if (err.ssl == SSL_ERROR_WANT_READ) {
```
I'd suggest using a switch here as well.

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2470,171 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+{
+#ifdef BIO_get_ktls_send
+    int uses = BIO_get_ktls_send(SSL_get_wbio(self->ssl));
+    return PyBool_FromLong((long)uses);
```
```suggestion
    return PyBool_FromLong(uses);
```

Promotion will be implicit.

picnixz reviewed (commented):
Some comments. The `if-else if` blocks need to follow PEP-7 but you can also use a switch if you want (it can be clearer to read).

illia-v committed:
Apply some suggestions from code review

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

illia-v committed:
Apply suggestions to Python code

illia-v committed:
Improve style of `_ssl__SSLSocket_sendfile_impl`

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Drop `_sendfile_use_ssl_sendfile`

illia-v committed:
Merge remote-tracking branch 'python/main' into SSL_sendfile

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1106,6 +1107,11 @@ SSL Sockets
       zero-length data no longer fails with a protocol violation error.
 
+   .. versionchanged:: next
+      Python now uses ``SSL_sendifle`` internally when it is possible. The
```
```suggestion
      Python now uses ``SSL_sendfile`` internally when possible. The
```

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -4297,19 +4297,30 @@ def test_read_write_after_close_raises_valuerror(self):
-            with client_context.wrap_socket(socket.socket(),
-                                            server_hostname=hostname) as s:
-                s.connect((HOST, server.port))
+        # kTLS seems to work only with a connection created before
```
Can you check whether it "seems" to work or not? (for a personal project I would accept this but I'm interested in knowing whether this is an issue with Python SSL or OpenSSL)

illia-v replied:
Okay, I'll post more details about this soon

illia-v replied:
I can reproduce the same issue with C code using OpenSSL (without Python). Please check my findings [here](https://skillful-cloche-642.notion.site/Discrepancy-between-wrapping-connected-and-unconnected-sockets-1daaaec11976800f853ae2f769dc3c5c). I mentioned a possible patch, let me know if we can apply it or there can be undesirable consequences.

FYI, `ssl.OP_ENABLE_KTLS` to enable kTLS has been available since Python 3.12 (and it could be added to options as a simple integer even before). And the issue is not specific to `SSL_sendfile`, it affects kTLS availability for simple reads and writes.

I added my reproducer to an existing OpenSSL issue https://github.com/openssl/openssl/issues/19676#issuecomment-2816750061.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -4297,19 +4297,30 @@ def test_read_write_after_close_raises_valuerror(self):
+            ) as ssock:
+                if support.verbose:
+                    ktls_used = ssock._sslobj.uses_ktls_for_send()
+                    print(
+                        'kTLS is',
+                        'available' if ktls_used else 'unavailable',
+                    )
```
Instead of using a print(), let's use `with self.subTest(ktls_used=ktls_used):`

illia-v replied:
If I use a sub-test and run `./python -m test -vv test_ssl`, the info is absent from output.

Can we leave the `print` call? Other tests in the file use `print` or `sys.stdout.write` too and the typical output looks like this:
![image](https://github.com/user-attachments/assets/fda687cf-e63f-4489-bd9e-bac716ed55cd)

picnixz replied:
Oh so it's really only for `-vv` and not just for `-v`. Ok for leaving the print. In general, I think it's better to only report information upon *failures* (which is why I suggested using subTest()), but with `-vv`, it's maybe fine.

illia-v replied:
The info is printed with a single `-v` too

picnixz replied:
If other tests are already printing information, I'm ok with printing something here.

picnixz commented on Misc/NEWS.d/next/Library/2023-03-13-22-51-40.gh-issue-99813.40TV02.rst:
```diff
@@ -0,0 +1,4 @@
+Python now uses ``SSL_sendfile`` internally when it is possible (see
+:data:`~ssl.OP_ENABLE_KTLS`.) The function sends a file more efficiently
+because it performs TLS encryption in the kernel to avoid additional context
+switches. Patch by Illia Volochii.
```
```suggestion
Python now uses ``SSL_sendfile`` internally when it is possible (see
:data:`~ssl.OP_ENABLE_KTLS`). The function sends a file more efficiently
because it performs TLS encryption in the kernel to avoid additional context
switches. Patch by Illia Volochii.
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -74,6 +74,33 @@
+    typedef long long Py_off_t;
+#else
+    typedef off_t Py_off_t;
+#endif
```
```suggestion
#ifdef BIO_get_ktls_send
#  ifdef MS_WINDOWS
typedef long long Py_off_t;
#  else
typedef off_t Py_off_t;
#  endif
```

Generally I think we indent the preprocessor directives, not the code inside.

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2467,181 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    if (sock != NULL) {
+        /* just in case the blocking state of the socket has been changed */
+        int nonblocking = (sock->sock_timeout >= 0);
+        BIO_set_nbio(SSL_get_rbio(self->ssl), nonblocking);
+        BIO_set_nbio(SSL_get_wbio(self->ssl), nonblocking);
```
Any possibility those calls fail or not?

illia-v replied:
Same lines have been present in `_ssl__SSLSocket_write_impl` and a few other functions for long time, I believe they are proof tested.

If there is possibility, can we consider it as a separate issue for the ssl module?

picnixz replied:
Actually from the manpage:

> BIO_set_nbio() always returns 1. 

So there isn't a way to make it fail. The only possibility is when `SSL_get_wbio` returns NULL, but I think the whole call becomes a no-op then.

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2467,181 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+        }
+
+        if (has_timeout) {
+            timeout = _PyDeadline_Get(deadline);
```
Is it possible that this one is negative? if so, shouldn't we break before?

illia-v replied:
There are four existent occurrences of these lines in the ssl module too. They were added by https://github.com/python/cpython/commit/833fdf126c8fe77fd17e8a8ffbc5c571b3bf64bd#diff-89879be484d86da4e77c90d5408b2e10190ee27cc86337cd0f8efc3520d60621

`PySSL_select` exits early with `SOCKET_HAS_TIMED_OUT` in case of negative values, the returned value is handled a few lines below

picnixz replied:
Ah I see, it's being handled in `PySSL_select`. Thanks.

picnixz reviewed (commented):
few final comments

illia-v committed:
Apply suggestions from code review

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Improve kTLS checks for older OpenSSL

https://github.com/openssl/openssl/commit/524bac570702a79366b85ff1f66e07d3e002370c

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2467,195 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+        Py_RETURN_FALSE;
+    }
+#else
+    Py_RETURN_FALSE;
```
Alternatively, how about:

```c
return Py_NewRef(uses == 1 ? Py_True : Py_False);
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2467,195 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+        }
+        else if (sockstate == SOCKET_HAS_BEEN_CLOSED) {
+            PyErr_SetString(get_state_sock(self)->PySSLErrorObject,
+                            "Underlying socket has been closed.");
```
Just checking this:

- if the error message already exists, use the same text (namely capitalized text ending with a period)
- if not, try to see if error messages are using periods or not and/or are capitalized and pick the one that is the most common

illia-v replied:
The exact error message exists in four other places in the file

illia-v replied:
Regarding the message "The sendfile operation timed out" above without the period, it's siblings don't end with a period too:
* "The handshake operation timed out"
* "The write operation timed out"
* "The read operation timed out"

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2467,195 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+            break;
+        }
+    } while (err.ssl == SSL_ERROR_WANT_READ ||
+             err.ssl == SSL_ERROR_WANT_WRITE);
```
```suggestion
    } while (err.ssl == SSL_ERROR_WANT_READ
             || err.ssl == SSL_ERROR_WANT_WRITE);
```

I missed the PEP-7 here (but I honestly find this ugly but well.. PEP-7 it is)

picnixz commented on Modules/_ssl.c:
```diff
@@ -2440,6 +2467,195 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+}
+
+/*[clinic input]
+_ssl._SSLSocket.uses_ktls_for_read
```
I didn't spot this but I think it's perferrable to call it `for_recv` instead of `for_read`

picnixz reviewed (commented):
I feel we need explicit functions for common error messages just to always use the same ones (but it's a separate issue)

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Rename `uses_ktls_for_read` to `uses_ktls_for_recv`

illia-v committed:
Apply PEP 7

illia-v committed:
Use an alternative method of returning booleans

picnixz commented on Modules/_ssl.c:
```diff
@@ -74,6 +74,33 @@
 
+#ifdef BIO_get_ktls_send
+#  ifdef MS_WINDOWS
+typedef long long Py_off_t;
```
I think I've seen this construction elsewhere. This is off topic but I think we should add it to the default clinic converters or be able to share code like that across clinic generated code. I need to think about it later.

picnixz commented on Modules/_ssl.c:
```diff
@@ -2441,6 +2468,185 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    // Also, it returns -1 for failure before OpenSSL 3.0.4.
+    return Py_NewRef(uses == 1 ? Py_True : Py_False);
+#else
+    return Py_NewRef(Py_False);
```
For this one, use Py_RETURN_FALSE

picnixz reviewed (approved):
I'd like @ZeroIntensity to have a look for threads safety and @gpshead to confirm the implementation as well. We could add a What's New entry but it's more of an implementation detail so I would prefer not to in the first place.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2441,6 +2468,185 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLSocket_uses_ktls_for_send_impl(PySSLSocket *self)
```
This function needs a critical section. Concurrent calls to `SSL_get_wbio` aren't thread-safe without the lock.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2441,6 +2468,185 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    int uses = BIO_get_ktls_send(SSL_get_wbio(self->ssl));
+    // BIO_get_ktls_send() returns 1 if kTLS is used and 0 if not.
+    // Also, it returns -1 for failure before OpenSSL 3.0.4.
+    return Py_NewRef(uses == 1 ? Py_True : Py_False);
```
This is an immortal reference, you can omit `Py_NewRef`.

picnixz replied:
Ideally it's better to make it explicit. We had a similar discussion on the discord and on some issues

ZeroIntensity replied:
Oh? I thought the consensus was to omit immortal reference counting.

picnixz replied:
I think there was no clear consensus but I think it's better to use the macro when possible and use Py_NewRef otherwise, even if the macro does not use Py_NewRef. IIRC, the discussion involved Irit and Serhiy so if you can find it and see what we eventually decided, it would be great! (I don't have the time nor the resources for that). And if I incorrectly remembered the consensus then I apologise in advance!

illia-v replied:
It started with returning `Py_False` without the wrapper as it looked like the most clean option, than I was asked to use `Py_NewRef` and `Py_RETURN_FALSE`.
"There should be one-- and preferably only one --obvious way to do it." is not applicable to returning booleans in CPython's C code definitely 😄

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2441,6 +2468,185 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLSocket_uses_ktls_for_recv_impl(PySSLSocket *self)
```
Ditto, let's add a critical section.

ZeroIntensity commented on Misc/NEWS.d/next/Library/2023-03-13-22-51-40.gh-issue-99813.40TV02.rst:
```diff
@@ -0,0 +1,4 @@
+Python now uses ``SSL_sendfile`` internally when it is possible (see
```
We know it's Python, but we don't know it's SSL:
```suggestion
:mod:`ssl` now uses ``SSL_sendfile`` internally when it is possible (see
```

picnixz replied:
It's the same wording as the previous entry (in the docs)

ZeroIntensity replied:
I think it makes sense in the docs because there's context. It's not immediately clear that we're talking about `ssl` here.

picnixz replied:
Oh right! let's do it your way

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2441,6 +2468,185 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+    int has_timeout;
+
+    if (sock != NULL) {
+        if ((PyObject *)sock == Py_None) {
```
Not a real issue, but thinking out loud for me and Bénédikt: does `Py_Is` come with `_PyObject_CAST`? If so, it seems [more useful](https://discuss.python.org/t/should-py-is-be-preferred-to/74323) than `==` here. If not, we should add it.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2441,6 +2468,185 @@ PySSL_select(PySocketSockObject *s, int writing, PyTime_t timeout)
+        Py_INCREF(sock);
+    }
+
+    if (sock != NULL) {
```
Why isn't this merged with the case above?

illia-v replied:
It was copied from implementation of the write method as is, I merged the two blocks now

ZeroIntensity commented on Doc/library/ssl.rst:
```diff
@@ -1071,8 +1071,9 @@ SSL Sockets
-   - :meth:`~socket.socket.sendfile` (but :mod:`os.sendfile` will be used
-     for plain-text sockets only, else :meth:`~socket.socket.send` will be used)
+   - :meth:`~socket.socket.sendfile` (but it may be high-performant only when
+     the kernel TLS is enabled (see :data:`~ssl.OP_ENABLE_KTLS`) or when a
```
I'm not sure what we do about nested parentheses in the docs. I haven't seen them anywhere else, so I'd like to avoid them here. Could we reword this a little to avoid that?

illia-v replied:
Reworded them ✅

illia-v committed:
Mark new functions with `critical_section`

illia-v committed:
Merge two if blocks

illia-v committed:
Reword docs

illia-v committed:
Update the news entry

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

illia-v committed:
Use `Py_RETURN_FALSE` again

illia-v committed:
Merge remote-tracking branch 'python/main' into SSL_sendfile

ZeroIntensity reviewed (approved):
I forgot to approve this, LGTM as well.

illia-v committed:
Merge branch 'main' into SSL_sendfile

illia-v committed:
Merge branch 'main' into SSL_sendfile

picnixz committed:
Merge branch 'main' into SSL_sendfile
